### ISSUE TITLE
Clarification needed on rand_hc implementation

### ISSUE THREAD

**Issue by m0xsec** [rust-random/rngs#55]
### Background:
The latest version of the [docs](https://docs.rs/rand_hc/latest/rand_hc/struct.Hc128Rng.html#fn4) indicate the following:

> This implementation is not based on the version of HC-128 submitted to the eSTREAM contest, but on a later version by the author with a few small improvements from December 15, 2009[4](https://docs.rs/rand_hc/latest/rand_hc/struct.Hc128Rng.html#fn4).

The footnote in the docs link to this reference that no longer exists - https://www.ntu.edu.sg/home/wuhj/research/hc/index.html

However, using the Internet Archive, we can see the page, with the last modified date of December 15, 2009 - https://web.archive.org/web/20110527154230/https://www.ntu.edu.sg/home/wuhj/research/hc/index.html

This webpage is general and does not directly outline said "small improvements" by the author of HC-128.

### Concern:
The current docs do not transparently reflect which version of HC-128 is implemented in `rand_rc`. The [Rust Random book](https://rust-random.github.io/book/guide-rngs.html) specifically mentions that HC-128 is recommended by eSTREAM, though this crate is using an implementation that is NOT recommended by eSTREAM.

This creates a false sense of security for those who may be relying on the guarantees and analysis that the eSTREAM project provided. eSTREAM did not provide analysis or review of the "small improvements" version that is vaguely referenced in the docs for `rand_rc`. 

Because of this, the docs need to better reflect which implementation is being used and **MUST** be very clear that the recommendation by eSTREAM is misleading.

**Comment by vks** [comment]
I think we can just remove HC-128 from the table, we are not using or recommending it anymore.

**Comment by m0xsec** [comment]
> I think we can just remove HC-128 from the table, we are not using or recommending it anymore.

Would this change just apply to the docs or are you suggesting removing HC-128 from the `rand_rc` codebase as well? Also, I am curious why its not being recommended anymore - any info you can point me to regarding that change?

**Comment by dhardy** [comment]
Only that we no longer use it as `StdRng`, so no longer need it in that sense.

I haven't looked into this in a long while, but it may be useful to continue maintaining this? That said, assistance with maintainance is always appreciated.

**Comment by m0xsec** [comment]
I would advocate for maintaining HC-128 as I believe it is still relevant. I do think this requires a review of the documentation and the implementation so it is clear what version of HC-128 is being provided. I can take a stab at trying to piece that together if it isn't readily available given how long its been.

**Comment by dhardy** [comment]
> I can take a stab at trying to piece that together if it isn't readily available given how long its been.

Please do!

**Comment by vks** [comment]
I went down the rabbit hole: The archived reference has two implementations: `hc128_ref.h` and `hc128_opt32.h`. The second implementation introduces some optimizations (loop unrolling, in-place key expansion), but does not change the cipher specified in Wu's paper, so the eSTREAM cryptanalysis applies to both. `rand_hc`'s implementation is clearly based on `hc128_opt.h`, with some minor changes.

All three implementations should produce identical keystreams. [We also test compliance using the official eSTREAM test vectors.](https://github.com/rust-random/rngs/blob/be9aa09cdfec726ccbc3acc7a01a59629ee46760/rand_hc/src/hc128.rs#L358-L416)

I think the only action item for this issue is to clarify in the docs what "small improvements" means.

**Comment by dhardy** [comment]
I agree with @vks's assessment: this implementation is able to reproduce the same test vectors, therefore it is compatible with the algorithm as published by Hongjun Wu for the eSTREAM contest. There are two limitations here:

- Our implementation uses a different API, for example it uses a 256-bit key instead of the 128-bit key + 128-bit IV used by the original publication. We believe the implementation to be compatible with the API and its security requirements, but of course the eSTREAM contest did not verify this.
- Our implementation is a translation to Rust, not a straight copy of the original code, thus it is not the code reviewed by the eSTREAM contest (regardless of which version of the C code it was based on).
